src/databases/MOAB/avtMOABFileFormat.C

```diff
@@ -86,26 +86,35 @@ avtMOABFileFormat::avtMOABFileFormat(const char *filename, const DBOptionsAttrib
 //
 // ****************************************************************************
 
-avtMOABFileFormat::~avtMOABFileFormat()
-{
-	debug1 << " avtMOABFileFormat::~avtMOABFileFormat: freeing file descriptor\n";
-	if (file_descriptor)
-	{
-		free(file_descriptor);
-		file_descriptor=NULL;
-	}
-	delete mbCore;
+avtMOABFileFormat::~avtMOABFileFormat() {
+    debug1 << " avtMOABFileFormat::~avtMOABFileFormat: freeing file descriptor\n";
+    if (file_descriptor) {
+        free(file_descriptor);
+        file_descriptor = NULL;
+    }
+    if (mbCore) {
+        delete mbCore;
+        mbCore = NULL;
+    }
 }
 void
-avtMOABFileFormat::FreeUpResources(void)
-{
-	debug1 << " avtMOABFileFormat::FreeUpResources: freeing file descriptor\n";
-    free  (file_descriptor);
-    file_descriptor = NULL;
+avtMOABFileFormat::FreeUpResources(void) {
+    debug1 << " avtMOABFileFormat::FreeUpResources: freeing file descriptor\n";
+    if (file_descriptor) {
+        free(file_descriptor);
+        file_descriptor = NULL;
+    }
 #ifdef PARALLEL
-    delete pcomm;
+	if (pcomm)
```

Comment:
Except for this `#ifdef PARALLEL` code block, the code here and in `avtMOABFileFormat::~avtMAOBFileFormat()` are identical. So, it begs the question why isn't the parallel code block handled in the destructor also?

And, it looks like the approach you've taken is to protect calls to `free()` with checks for non-`NULL` and to do it in both the destructor and in `FreeUpResources()` just to be safe. Thats ok.

You have a `STMD` database (single-time step, multiple domain)...which means that VisIt thinks that the on-disk *files* contain just one timestep but may contain multiple domains. When VisIt needs to change time steps, it calls `FreeUpResources` which is the plugin's cue to free resources associated with the current timestep. Normally, that would be closing the file(s) associated with the current timestep but not necessarily destroying stuff that is needed from one timestep to the next (I am thinking the `pcomm` object for example). If that object should persist across time steps, then don't free it in `FreeUpResources` (unless you plan to re-create it upon the next timestep *and* that is a low cost operation to perform).

Does this make sense?

Reply:
Thanks @markcmiller86  

We can also have a list of files, one per time step. How can we use the player then to animate? (When I open vtk files that are numbered, I have the option to open "smart" , and they are playing in animation).  I do not have that option for h5m files, and the VTK player is also STMD.
 
how can we achieve that ?

Reply:
@iulian787 I am returning to this PR and am a bit confused by several things now. First, I assume by "player", you mean VisIt. Is that right? Next the notion of opening files "smart"ly has to do with whether VisIt recognizes a sequence of files on disk as a sequence of timesteps. But, for that to work, the files have to follow some restricted convention(s) in the way they are named. If that doesn't work for your files, you can always resort to creating a `.visit` file (which simply lists all the files in the time series in order).

And, this is all only very loosely related to my original question...what is your rational for freeing (or not freeing) the things you do in `FreeUpResources` and the destructor? The destructor is called when the database is closed. So, *everything* should ultimately be freed there. `FreeUpResources` is called when VisIt is changing timesteps. So, you should free up there only those resources needed to process the current timestep.

Reply:
It looks like you have two PRs in flight to do the same work on both the current release candidate and for develop. We really only want/need to review one of those and typically start with the PR to the RC. Once that is merged, then do a `cherry-pick` of that one merge commit on the RC to `develop` and assign me as reviewer. I will just blindly merge that second one because all the necessary review is completed in the one to the RC. OK?

Reply:
we have one file per time step, in MOAB, we do not have a convention for multiple time steps in a file. 
maybe I am confused what database means. so far, in my understanding, one file, one database. So if I need a new time step, I close the existing file, open a new one. (it has the same variables/tags names) 

So if I want to use player in visit to animate, I use mbconvert to convert to vtk files. Instead of step01.h5m, step02.h5m, etc, I have step01.vtk, step02.vtk. VisIt has the option to open smartly the vtk files, and animate them, but not h5m files. That is the mechanism I do not get how it works.  (h5m files and vtk files both have the same mesh, and the fields I need, let's say temperature).

Reply:
> we have one file per time step, in MOAB, we do not have a convention for multiple time steps in a file. maybe I am confused what database means. so far, in my understanding, one file, one database. 

In VisIt, a *database* is a bucket of states of one or more meshes (which may or may not also include *variables* defined on the mesh(s)) , typically organized in a series representing *time*. How states are organized across *files* is totally arbitrary. One file could store all states or you could have one file for each state or you could have a single state defined by a slew of files.

VisIt's four database interface types (`STSD`, `STMD`, `MTSD`, and `MTMD`) are really just a mechanism for making life simple for plugin developers who typically only need to worry about a single timestep (`ST`) and single domain (`SD`) mesh. They help to hide some of the file system complexity for the simple cases. In theory, all our plugins could just be `MTMD` type but then every plugin developer would have to engage in additional coding to manage more of the file system details (like making sure no more than the limit of file descriptors is open at any one time).

So if I need a new time step, I close the existing file, open a new one. (it has the same variables/tags names). STMD means I have one time step per database/file ?

Yes, that makes sense. But, are there other resources the MOAB plugin uses that *live* beyond any one file **AND** cost in terms of time to construct when needed? If, so, don't free those in `FreeUpResources` and instead free them in the destructor.

> 
> Right now, if I want to use player in VisIt to animate, I use mbconvert to convert to vtk files. Instead of step01.h5m, step02.h5m, etc, I have step01.vtk, step02.vtk. VisIt has the option to open smartly the vtk files, and animate them, but not h5m files. That is the mechanism I do not get how it works. (h5m files and vtk files both have the same mesh, and the fields I need, let's say temperature).

Ok, I think this is wholly unrelated to your plugin and instead has something to do with some internal logic in VisIt that is trying to group filenames. I will investigate shortly.

> 
> our pcomm object is created per file. Also, we have a new instance of moab for each file. Maybe that has to change, to create only one moab instance, when the plugin is activated. That would be a big change though, and make sure we are not leaking anything.

Ah, ok, so fine, delete `pcomm` in `FreeUpResources` too.

Reply:
> Ok, I think this is wholly unrelated to your plugin and instead has something to do with some internal logic in VisIt that is trying to group filenames. I will investigate shortly.

yeah, this is internal to VisIt in its file group logic probably around here...

https://github.com/visit-dav/visit/blob/d1fbae6caeb3364954f3c2b54eb0a27ae69fd889/src/mdserver/main/MDServerConnection.C#L1807

That is a really complex code block.

I can create this issue just by taking several vtk files and changing their extension...

![Screen Shot 2022-08-19 at 5 53 48 PM](https://user-images.githubusercontent.com/5720676/185723153-c2317165-9cd5-4ba8-bf77-184ded67dc98.png)

I'll file a ticket.